### PR TITLE
Detect the AMD TSME encryption state for HSI

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

File: libfwupdplugin/fu-common.c [hunk 1]
```diff
@@ -2258,6 +2258,34 @@ fu_common_is_cpu_intel (void)
...
+ * fu_common_is_cpu_amd:
+ *
+ * Uses CPUID to discover the CPU vendor and check if it is AMD.
+ *
+ * Return value: %TRUE if the vendor was AMD.
+ *
+ * Since: 1.5.5
+ **/
+gboolean
+fu_common_is_cpu_amd (void)
```
Comment (hughsie) [inline libfwupdplugin/fu-common.c:2271]: of course, this implies adding fu_common_is_cpu_riscv and fu_common_is_cpu_arm in the future. Maybe something like `fu_common_get_cpu_vendor()` might be better, with a `FuCpuVendor` enum with `FU_CPU_VENDOR_AMD` etc.

Comment (superm1) [reply libfwupdplugin/fu-common.c:2271]: I think that approach is better, a lot less symbols to maintain for ABI.

File: plugins/msr/fu-plugin-msr.c [hunk 2]
```diff
@@ -185,9 +223,41 @@ fu_plugin_add_security_attr_dci_locked (FuPlugin *plugin, FuSecurityAttrs *attrs
 	fwupd_security_attr_set_result (attr, FWUPD_SECURITY_ATTR_RESULT_LOCKED);
 }
 
+static void
+fu_plugin_add_security_attr_tsme_enabled (FuPlugin *plugin, FuSecurityAttrs *attrs)
+{
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	g_autoptr(FwupdSecurityAttr) attr = NULL;
+
+	/* this MSR is only valid for a subset of Intel CPUs */
```
Comment (superm1) [inline plugins/msr/fu-plugin-msr.c:232]: AMD

File: libfwupdplugin/fu-common.c (hunk 1 above)
Comment (hughsie) [reply libfwupdplugin/fu-common.c]: Agree; fixed.

File: plugins/msr/fu-plugin-msr.c (hunk 2 above)
Comment (hughsie) [reply plugins/msr/fu-plugin-msr.c]: Ta, fixed.

File: libfwupdplugin/fu-common.c [hunk 3]
```diff
@@ -2236,26 +2236,49 @@ fu_common_cpuid (guint32 leaf,
...
+{
+	return fu_common_get_cpu_vendor () == FU_CPU_VENDOR_INTEL;
+}
+
+/**
+ * fu_common_get_cpu_vendor:
+ *
+ * Uses CPUID to discover the CPU vendor.
+ *
+ * Return value: %TRUE if the vendor was AMD.
```
Comment (superm1) [inline libfwupdplugin/fu-common.c:2254]: wrong comment now

File: libfwupdplugin/fu-common.c [hunk 4]
```diff
@@ -2236,26 +2236,49 @@ fu_common_cpuid (guint32 leaf,
...
+FuCpuVendor
+fu_common_get_cpu_vendor (void)
 {
 	guint ebx = 0;
 	guint ecx = 0;
 	guint edx = 0;
 
-	if (!fu_common_cpuid (0x0, NULL, &ebx, &ecx, &edx, NULL))
-		return FALSE;
+	if (fu_common_cpuid (0x0, NULL, &ebx, &ecx, &edx, NULL)) {
```
Comment (superm1) [inline libfwupdplugin/fu-common.c:2265]: in the case of no `#ifdef` does the compiler let you do this?  I would think you need to have something in the clause.  IE switch it to 
```
if (!fu...)
   return FU_CPU_VENDOR_UNKNOWN;
#ifdef ...
//do stuff
#endif 
return  FU_CPU_VENDOR_UNKNOWN;
```

File: plugins/msr/fu-plugin-msr.c [hunk 5]
```diff
@@ -41,12 +53,23 @@ gboolean
...
 {
 	FuPluginData *priv = fu_plugin_get_data (plugin);
+	guint eax = 0;
 	guint ecx = 0;
 
 	/* sdbg is supported: https://en.wikipedia.org/wiki/CPUID */
-	if (!fu_common_cpuid (0x01, NULL, NULL, &ecx, NULL, error))
-		return FALSE;
-	priv->ia32_debug_supported = ((ecx >> 11) & 0x1) > 0;
+	if (fu_common_get_cpu_vendor () != FU_CPU_VENDOR_INTEL) {
```
Comment (superm1) [inline plugins/msr/fu-plugin-msr.c:60]: wait, don't you want to be checking **for** Intel?

Comment (hughsie) [reply plugins/msr/fu-plugin-msr.c:60]: Indeed, doh.

File: libfwupdplugin/fu-common.c (hunk 4 above)
Comment (hughsie) [reply libfwupdplugin/fu-common.c:2265]: The compiler seems happy enough with the empy if clause, although I agree it's a bit wacky. I'll just move the entire function minus the last lines into the ifdef.

File: libfwupdplugin/fu-common.c (hunk 3 above)
Comment (hughsie) [reply libfwupdplugin/fu-common.c:2254]: Fixed.

**Review by superm1 — APPROVED** [review]

**Comment by superm1** [comment]
Does HSI documentation need something added for AMD's TSME in the mentions?

**Comment by hughsie** [comment]
> Does HSI documentation need something added for AMD's TSME in the mentions?

Yup! Will fix and merge, thanks.

**Pull request merged**